src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+    } else if (header.hashPrevBlock != m_redownload_buffer_last_hash) {
+        return false;
+    }
+
+    int64_t next_height = m_redownload_buffer_last_height + 1;
+
+    // If we're at a header for which we previously stored a commitment, verify
+    // it is correct. Failure will result in aborting download.
+    if ((next_height - m_chain_start->nHeight) % HEADER_COMMITMENT_FREQUENCY == 0) {
+         bool commitment = m_hasher(header.GetHash()) & 1;
+         if (m_header_commitments.size() == 0) {
+            // Somehow our peer managed to feed us a different chain and
+            // we've run out of commitments.
+            return false;
+        }
+        bool expected_commitment = m_header_commitments.front();
+        m_header_commitments.pop_front();
+        if (commitment != expected_commitment) {
+            return false;
+        }
+    }
+
+    // Store this header for later processing.
+    m_redownloaded_headers.push_back(header);
+    m_redownload_buffer_last_height = next_height;
+    m_redownload_buffer_last_hash = header.GetHash();
+
+    // If we're processing our target block header, which we verified has
+    // sufficient work, then set a flag for processing all remaining headers.
+    if (header.GetHash() == m_blockhash_with_sufficient_work) {
+        m_process_all_remaining_headers = true;
+    }
+    return true;
+}
+
+std::vector<CBlockHeader> HeadersSyncState::RemoveHeadersReadyForAcceptance()
+{
+    std::vector<CBlockHeader> ret;
+
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return ret;
+
+    while (m_redownloaded_headers.size() > REDOWNLOAD_HEADERS_THRESHOLD ||
+            (m_redownloaded_headers.size() > 0 && m_process_all_remaining_headers)) {
+        ret.emplace_back(m_redownloaded_headers.front().GetFullHeader(m_redownload_buffer_first_prev_hash));
+        m_redownloaded_headers.pop_front();
+        m_redownload_buffer_first_prev_hash = ret.back().GetHash();
+    }
+    return ret;
+}
+
+std::optional<CBlockLocator> HeadersSyncState::MakeNextHeadersRequest()
+{
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return std::nullopt;
+
+    std::vector<uint256> locator;
+
+    if (m_download_state == State::INITIAL_DOWNLOAD && !m_last_header_received.IsNull()) {
+        // During initial download, we continue from the last header received.
+        locator.push_back(m_last_header_received.GetHash());
+    }
+
+    if (m_download_state == State::REDOWNLOAD && !m_redownloaded_headers.empty()) {
+        // During redownload, we will either download from the last received
+        // header that we stored during the second download phase, or from the
+        // fork point (m_chain_start).
+        locator.push_back(m_redownload_buffer_last_hash);
+    }
+
+    locator.insert(locator.end(), m_chain_start_locator.vHave.begin(),
+            m_chain_start_locator.vHave.end());
+    return CBlockLocator(locator);
```

Comment:
In commit "Utilize anti-DoS headers download strategy"

`return CBlockLocator(std::move(locator))` saves a copy.

Reply:
Done.